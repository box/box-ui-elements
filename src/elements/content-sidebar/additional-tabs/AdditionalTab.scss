@import '../../../styles/variables';
@import '../../../styles/common/_animations';
@import '../mixins';

.bdl-AdditionalTab {
    @include bdl-SidebarNavButton;

    &:hover {
        svg path {
            fill: $bdl-gray-80;
        }
    }
}

.bdl-AdditionalTab-icon {
    @include bdl-openComponentAnimation(.5s);
    @include bdl-SidebarNavIcon;
}
