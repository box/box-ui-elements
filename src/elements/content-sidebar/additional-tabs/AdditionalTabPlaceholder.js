/**
 * @flow
 * @file Preview sidebar additional tab loading component
 * @author Box
 */

import classNames from 'classnames';
import * as React from 'react';

type Props = {
    isLoading: boolean,
};

const AdditionalTabPlaceholder = ({ isLoading = false }: Props) => {
    const classes = classNames('bdl-AdditionalTab-icon', 'bdl-AdditionalTabLoading-placeholder', {
        'bdl-AdditionalTabLoading-placeholder--loading': isLoading,
    });
    return (
        <div className="bdl-AdditionalTabLoading">
            <div className={classes} />
        </div>
    );
};

export default AdditionalTabPlaceholder;
