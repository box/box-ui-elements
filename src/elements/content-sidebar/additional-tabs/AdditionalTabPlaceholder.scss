.bdl-AdditionalTabPlaceholder {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 40px;
    height: 40px;

    &.bdl-is-loading {
        background-color: #e8e8e8;
        border-radius: 4px;
    }

    &-icon {
        display: flex;
        align-items: center;
        justify-content: center;
    }
}
