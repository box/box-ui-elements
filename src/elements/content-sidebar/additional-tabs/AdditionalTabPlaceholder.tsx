import * as React from 'react';

interface AdditionalTabPlaceholderProps {
    isLoading: boolean;
}

const AdditionalTabPlaceholder = ({ isLoading }: AdditionalTabPlaceholderProps) => (
    <div className={`bdl-AdditionalTabPlaceholder ${isLoading ? 'bdl-is-loading' : ''}`} />
);

export default AdditionalTabPlaceholder;
