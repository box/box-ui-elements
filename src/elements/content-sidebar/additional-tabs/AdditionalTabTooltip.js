/**
 * @flow
 * @file Sidebar Additional Tab FTUX tooltip
 * @author Box
 */

import * as React from 'react';
import Tooltip from '../../common/Tooltip';
import TargetedClickThroughGuideTooltip from '../../../features/targeting/TargetedClickThroughGuideTooltip';
import type { AdditionalSidebarTabFtuxData } from '../flowTypes';
import './AdditionalTabTooltip.scss';

type Props = {
    children: React.Node,
    defaultTooltipText: ?string | React.Node,
    ftuxTooltipData?: AdditionalSidebarTabFtuxData,
    isFtuxVisible: boolean,
};

const AdditionalTabTooltip = ({ children, defaultTooltipText, isFtuxVisible, ftuxTooltipData }: Props) => {
    if (!isFtuxVisible || !ftuxTooltipData || !ftuxTooltipData.targetingApi().canShow) {
        return (
            <Tooltip position="middle-left" text={defaultTooltipText}>
                {children}
            </Tooltip>
        );
    }

    const { targetingApi, text } = ftuxTooltipData;

    return (
        <TargetedClickThroughGuideTooltip
            className="bdl-AdditionalTabTooltip"
            body={text}
            position="middle-right"
            shouldTarget
            useTargetingApi={targetingApi}
        >
            {children}
        </TargetedClickThroughGuideTooltip>
    );
};

export default AdditionalTabTooltip;
