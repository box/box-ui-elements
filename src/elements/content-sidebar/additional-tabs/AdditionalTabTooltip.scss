.bdl-AdditionalTabTooltip.tooltip {
    // Max-width is chosen so the tooltip is small enough to fit on the right-hand side of
    // the sidebar. Using the default max-width forces the tooltip to the left-hand side.
    max-width: 250px;
}
