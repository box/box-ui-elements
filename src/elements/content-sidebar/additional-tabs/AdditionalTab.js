/**
 * @flow
 * @file Sidebar Additional Tab component
 * @author Box
 */

import * as React from 'react';
import classNames from 'classnames';
import { FormattedMessage } from 'react-intl';

import { bdlGray50 } from '../../../styles/variables';
import PlainButton from '../../../components/plain-button/PlainButton';
import IconEllipsis from '../../../icons/general/IconEllipsis';
import AdditionalTabTooltip from './AdditionalTabTooltip';
import AdditionalTabPlaceholder from './AdditionalTabPlaceholder';
import messages from './messages';
import type { AdditionalSidebarTab, AdditionalSidebarTabFtuxData } from '../flowTypes';

import './AdditionalTab.scss';

type Props = {
    ftuxTooltipData?: AdditionalSidebarTabFtuxData,
    isLoading: boolean,
    onImageLoad: () => void,
    status?: string,
} & AdditionalSidebarTab;

type State = {
    isErrored: boolean,
};

const BLOCKED_BY_SHEILD = 'BLOCKED_BY_SHIELD_ACCESS_POLICY';

class AdditionalTab extends React.PureComponent<Props, State> {
    state = {
        isErrored: false,
    };

    onImageError = () => {
        this.props.onImageLoad();
        this.setState({ isErrored: true });
    };

    isDisabled() {
        const { status } = this.props;
        return status === BLOCKED_BY_SHEILD;
    }

    getDisabledReason() {
        let reason = '';
        const { status } = this.props;
        switch (status) {
            case BLOCKED_BY_SHEILD:
                reason = <FormattedMessage {...messages.blockedByShieldAccessPolicy} />;
                break;
            default:
            // noop
        }
        return reason;
    }

    getTabIcon() {
        const { id, iconUrl, onImageLoad, title } = this.props;
        const { isErrored } = this.state;

        let TabIcon;

        if (isErrored) {
            TabIcon = <AdditionalTabPlaceholder isLoading={false} />;
        } else if (id && id > 0) {
            TabIcon = (
                <img
                    className="bdl-AdditionalTab-icon"
                    src={iconUrl}
                    onError={this.onImageError}
                    onLoad={onImageLoad}
                    alt={title}
                />
            );
        } else {
            TabIcon = <IconEllipsis color={bdlGray50} />;
        }

        return TabIcon;
    }

    render() {
        const {
            callback: callbackFn,
            id,
            isLoading,
            iconUrl,
            ftuxTooltipData,
            onImageLoad,
            title,
            ...rest
        } = this.props;

        const isDisabled = this.isDisabled();

        const className = classNames('bdl-AdditionalTab', {
            'bdl-is-hidden': isLoading,
            'bdl-is-disabled': isDisabled,
        });

        const tooltipText = isDisabled ? this.getDisabledReason() : title;

        return (
            <AdditionalTabTooltip
                defaultTooltipText={tooltipText}
                ftuxTooltipData={ftuxTooltipData}
                isFtuxVisible={!isLoading}
            >
                <PlainButton
                    aria-label={title}
                    className={className}
                    data-testid="additionaltab"
                    type="button"
                    isDisabled={isDisabled}
                    onClick={() => callbackFn({ id, callbackData: rest })}
                >
                    {this.getTabIcon()}
                </PlainButton>
            </AdditionalTabTooltip>
        );
    }
}

export default AdditionalTab;
