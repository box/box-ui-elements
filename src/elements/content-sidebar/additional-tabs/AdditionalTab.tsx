import * as React from 'react';
import classNames from 'classnames';
import { FormattedMessage } from 'react-intl';
import Apps16 from '../../../icon/fill/Apps16';

import PlainButton from '../../../components/plain-button/PlainButton';
import AdditionalTabTooltip from './AdditionalTabTooltip';
import AdditionalTabPlaceholder from './AdditionalTabPlaceholder';
import messages from '../../common/messages';
import type { AdditionalSidebarTab, AdditionalSidebarTabFtuxData } from '../flowTypes';
import './AdditionalTab.scss';

interface Props extends AdditionalSidebarTab {
    ftuxTooltipData?: AdditionalSidebarTabFtuxData;
    isLoading: boolean;
    onImageLoad: () => void;
    status?: string;
}

const BLOCKED_BY_SHIELD = 'BLOCKED_BY_SHIELD_ACCESS_POLICY';

const AdditionalTab = ({
    callback: callbackFn,
    iconUrl,
    id,
    isLoading,
    onImageLoad,
    title,
    icon,
    ftuxTooltipData,
    status,
    ...rest
}: Props): React.ReactElement => {
    const [isErrored, setIsErrored] = React.useState<boolean>(false);

    const onImageError = (): void => {
        onImageLoad();
        setIsErrored(true);
    };

    const isDisabled = (): boolean => {
        return status === BLOCKED_BY_SHIELD;
    };

    const getDisabledReason = (): React.ReactNode => {
        let reason: React.ReactNode = null;
        switch (status) {
            case BLOCKED_BY_SHIELD:
                reason = <FormattedMessage {...messages.blockedByShieldAccessPolicy} />;
                break;
            default:
            // noop
        }
        return reason;
    };

    const getTabIcon = (): React.ReactNode => {
        let TabIcon;

        if (isErrored) {
            TabIcon = <AdditionalTabPlaceholder isLoading={false} />;
        } else if (id && id > 0) {
            TabIcon = (
                <img
                    className="bdl-AdditionalTab-icon"
                    src={iconUrl}
                    onError={onImageError}
                    onLoad={onImageLoad}
                    alt={title}
                />
            );
        } else {
            TabIcon = icon || <Apps16 className="bdl-AdditionalTab-icon" />;
        }

        return TabIcon;
    };

    const disabled = isDisabled();

    const className = classNames('bdl-AdditionalTab', {
        'bdl-is-hidden': isLoading,
        'bdl-is-disabled': disabled,
        'bdl-is-overflow': id && id < 0,
    });

    const tooltipText = disabled ? getDisabledReason() : title;

    return (
        <AdditionalTabTooltip
            defaultTooltipText={tooltipText}
            ftuxTooltipData={ftuxTooltipData}
            isFtuxVisible={!isLoading}
        >
            <PlainButton
                aria-label={title}
                className={className}
                data-testid="additionaltab"
                isDisabled={disabled}
                onClick={() => callbackFn({ id, callbackData: rest })}
            >
                {getTabIcon()}
            </PlainButton>
        </AdditionalTabTooltip>
    );
};

export default AdditionalTab;
