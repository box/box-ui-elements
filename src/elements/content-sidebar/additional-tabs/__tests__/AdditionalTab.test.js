import React from 'react';
import { shallow } from 'enzyme';
import PlainButton from '../../../../components/plain-button/PlainButton';
import AdditionalTab from '../AdditionalTab';
import AdditionalTabTooltip from '../AdditionalTabTooltip';
import AdditionalTabPlaceholder from '../AdditionalTabPlaceholder';

describe('elements/content-sidebar/additional-tabs/AdditionalTab', () => {
    const getWrapper = props => shallow(<AdditionalTab {...props} />);

    test('should render the tooltip and button contents', () => {
        const mockSrc = 'https://foo.com/image';
        const props = {
            title: 'test title',
            iconUrl: mockSrc,
            id: 4,
            callback: () => {},
        };

        const wrapper = getWrapper(props);

        expect(
            wrapper
                .find(PlainButton)
                .childAt(0)
                .prop('src'),
        ).toEqual(mockSrc);
        expect(wrapper.find(AdditionalTabTooltip).prop('defaultTooltipText')).toBe('test title');

        expect(wrapper).toMatchSnapshot();
    });

    test('should render the more icon ellipsis if no valid id and icon URL are provided', () => {
        const props = {
            title: 'test title',
            id: -1,
            callback: () => {},
        };

        const wrapper = getWrapper(props);
        expect(wrapper).toMatchSnapshot();
    });

    test('should render the placeholder when an error occurs', () => {
        const props = {
            title: 'test title',
            id: -1,
            callback: () => {},
        };

        const wrapper = getWrapper(props);

        wrapper.setState({ isErrored: true });

        expect(wrapper.find(AdditionalTabPlaceholder)).toHaveLength(1);
        expect(wrapper).toMatchSnapshot();
    });

    test('should render disabled button when blocked by shield access policy', () => {
        const mockSrc = 'https://foo.com/image';
        const props = {
            title: 'test title',
            iconUrl: mockSrc,
            id: 4,
            callback: () => {},
            status: 'BLOCKED_BY_SHIELD_ACCESS_POLICY',
        };

        const wrapper = getWrapper(props);

        expect(wrapper).toMatchSnapshot();
    });

    test('should render the FTUX tooltip when ftuxTooltipData is present and the tab is not loading', () => {
        const mockSrc = 'https://foo.com/image';
        const props = {
            title: 'test title',
            iconUrl: mockSrc,
            id: 4,
            isLoading: false,
            ftuxTooltipData: {
                targetingApi: () => {},
                test: 'ftux tooltip text',
            },
            callback: () => {},
        };

        const wrapper = getWrapper(props);

        expect(wrapper.find(AdditionalTabTooltip).exists()).toBeTruthy();
    });
});
