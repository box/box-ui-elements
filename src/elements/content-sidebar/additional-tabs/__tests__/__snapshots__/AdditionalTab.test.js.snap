// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`elements/content-sidebar/additional-tabs/AdditionalTab should render disabled button when blocked by shield access policy 1`] = `
<AdditionalTabTooltip
  defaultTooltipText={
    <FormattedMessage
      defaultMessage="Use of this app is blocked due to a security policy."
      id="be.additionalTab.blockedByShieldAccessPolicy"
    />
  }
  isFtuxVisible={true}
>
  <PlainButton
    aria-label="test title"
    className="bdl-AdditionalTab bdl-is-disabled"
    data-testid="additionaltab"
    isDisabled={true}
    onClick={[Function]}
    type="button"
  >
    <img
      alt="test title"
      className="bdl-AdditionalTab-icon"
      onError={[Function]}
      src="https://foo.com/image"
    />
  </PlainButton>
</AdditionalTabTooltip>
`;

exports[`elements/content-sidebar/additional-tabs/AdditionalTab should render the more icon ellipsis if no valid id and icon URL are provided 1`] = `
<AdditionalTabTooltip
  defaultTooltipText="test title"
  isFtuxVisible={true}
>
  <PlainButton
    aria-label="test title"
    className="bdl-AdditionalTab"
    data-testid="additionaltab"
    isDisabled={false}
    onClick={[Function]}
    type="button"
  >
    <IconEllipsis
      color="#909090"
    />
  </PlainButton>
</AdditionalTabTooltip>
`;

exports[`elements/content-sidebar/additional-tabs/AdditionalTab should render the placeholder when an error occurs 1`] = `
<AdditionalTabTooltip
  defaultTooltipText="test title"
  isFtuxVisible={true}
>
  <PlainButton
    aria-label="test title"
    className="bdl-AdditionalTab"
    data-testid="additionaltab"
    isDisabled={false}
    onClick={[Function]}
    type="button"
  >
    <AdditionalTabPlaceholder
      isLoading={false}
    />
  </PlainButton>
</AdditionalTabTooltip>
`;

exports[`elements/content-sidebar/additional-tabs/AdditionalTab should render the tooltip and button contents 1`] = `
<AdditionalTabTooltip
  defaultTooltipText="test title"
  isFtuxVisible={true}
>
  <PlainButton
    aria-label="test title"
    className="bdl-AdditionalTab"
    data-testid="additionaltab"
    isDisabled={false}
    onClick={[Function]}
    type="button"
  >
    <img
      alt="test title"
      className="bdl-AdditionalTab-icon"
      onError={[Function]}
      src="https://foo.com/image"
    />
  </PlainButton>
</AdditionalTabTooltip>
`;
